controller-gen doesn't work with k8s.io/apiextensions-apiserver

Hi folks,

I use controller-gen integrated with Kubebuilder to generate RBAC, webhook and CRDs for my project, there is a command called `make manifests` to run:
 `controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases`   
to generate everything, but it doesn't work for me since I have `k8s.io/apiextensions-apiserver` in my `go.mod` as a dependency, the following error messages are poping out:
`/Users/maxwell/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.0.0-20190409022649-727a075fdec8/pkg/apis/apiextensions/v1beta1/types_jsonschema.go:63:2: enountered struct field "Raw" without JSON tag in type "JSON"
/Users/maxwell/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.0.0-20190409022649-727a075fdec8/pkg/apis/apiextensions/v1beta1/types_jsonschema.go:29:24: unsupported type "float64"
/Users/maxwell/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.0.0-20190409022649-727a075fdec8/pkg/apis/apiextensions/v1beta1/types_jsonschema.go:31:24: unsupported type "float64"
/Users/maxwell/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.0.0-20190409022649-727a075fdec8/pkg/apis/apiextensions/v1beta1/types_jsonschema.go:39:24: unsupported type "float64"
/Users/maxwell/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.0.0-20190409022649-727a075fdec8/pkg/apis/apiextensions/v1beta1/types_jsonschema.go:85:2: enountered struct field "Schema" without JSON tag in type "JSONSchemaPropsOrArray"
/Users/maxwell/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.0.0-20190409022649-727a075fdec8/pkg/apis/apiextensions/v1beta1/types_jsonschema.go:86:2: enountered struct field "JSONSchemas" without JSON tag in type "JSONSchemaPropsOrArray"
/Users/maxwell/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.0.0-20190409022649-727a075fdec8/pkg/apis/apiextensions/v1beta1/types_jsonschema.go:105:2: enountered struct field "Allows" without JSON tag in type "JSONSchemaPropsOrBool"
/Users/maxwell/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.0.0-20190409022649-727a075fdec8/pkg/apis/apiextensions/v1beta1/types_jsonschema.go:106:2: enountered struct field "Schema" without JSON tag in type "JSONSchemaPropsOrBool"
/Users/maxwell/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.0.0-20190409022649-727a075fdec8/pkg/apis/apiextensions/v1beta1/types_jsonschema.go:127:2: enountered struct field "Schema" without JSON tag in type "JSONSchemaPropsOrStringArray"
/Users/maxwell/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.0.0-20190409022649-727a075fdec8/pkg/apis/apiextensions/v1beta1/types_jsonschema.go:128:2: enountered struct field "Property" without JSON tag in type "JSONSchemaPropsOrStringArray"
Error: not all generators ran successfully
`
Any ideas about why this happened?

Thanks,
Jiacheng

Can you share your types.go file containing your API definitions? It looks like you are depending on something in the apiextensions-apiserver API group 🤔 fwiw, #277 is a fix for:

```
/Users/maxwell/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.0.0-20190409022649-727a075fdec8/pkg/apis/apiextensions/v1beta1/types_jsonschema.go:63:2: enountered struct field "Raw" without JSON tag in type "JSON"
```

We're still seeing this same error when using `controller-gen@v0.4.0`:

```
❯ controller-gen crd:trivialVersions=true paths=./pkg/...
/usr/local/google/home/juliankatz/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.18.2/pkg/apis/apiextensions/v1beta1/types_jsonschema.go:60:24: unsupported type "float64"
/usr/local/google/home/juliankatz/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.18.2/pkg/apis/apiextensions/v1beta1/types_jsonschema.go:62:24: unsupported type "float64"
/usr/local/google/home/juliankatz/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.18.2/pkg/apis/apiextensions/v1beta1/types_jsonschema.go:70:24: unsupported type "float64"
github.com/open-policy-agent/frameworks/constraint/pkg/apis/templates/v1alpha1:-: CRD for ConstraintTemplate.templates.gatekeeper.sh has no storage version
Error: not all generators ran successfully
run `controller-gen crd:trivialVersions=true paths=./pkg/... -w` to see all available markers, or `controller-gen crd:trivialVersions=true paths=./pkg/... -h` for usage
```

Just to be sure:

```
❯ controller-gen --version
Version: v0.4.0
```

Any other context we/I can provide to help get this addressed?  We're stuck on `v0.1.8` in https://github.com/open-policy-agent/frameworks/tree/master/constraint

Thanks Julian. Sadly that does not work for me:

```
/Users/tshak/code/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.18.12/pkg/apis/apiextensions/v1/types_jsonschema.go:191:2: encountered struct field "Schema" without JSON tag in type "JSONSchemaPropsOrArray"
/Users/tshak/code/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.18.12/pkg/apis/apiextensions/v1/types_jsonschema.go:192:2: encountered struct field "JSONSchemas" without JSON tag in type "JSONSchemaPropsOrArray"
/Users/tshak/code/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.18.12/pkg/apis/apiextensions/v1/types_jsonschema.go:211:2: encountered struct field "Allows" without JSON tag in type "JSONSchemaPropsOrBool"
/Users/tshak/code/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.18.12/pkg/apis/apiextensions/v1/types_jsonschema.go:212:2: encountered struct field "Schema" without JSON tag in type "JSONSchemaPropsOrBool"
/Users/tshak/code/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.18.12/pkg/apis/apiextensions/v1/types_jsonschema.go:233:2: encountered struct field "Schema" without JSON tag in type "JSONSchemaPropsOrStringArray"
/Users/tshak/code/go/pkg/mod/k8s.io/apiextensions-apiserver@v0.18.12/pkg/apis/apiextensions/v1/types_jsonschema.go:234:2: encountered struct field "Property" without JSON tag in type "JSONSchemaPropsOrStringArray"
```

I'm running `controller-gen` as part of go code generation:

```
//go:generate go run -tags generate sigs.k8s.io/controller-tools/cmd/controller-gen object:headerFile=../../hack/boilerplate.go.txt paths=./... crd:trivialVersions=true,crdVersions=v1,allowDangerousTypes=true output:artifacts:config=../../config/base
```

--- a/.golangci.yml
+++ b/.golangci.yml
@@ -1,4 +1,5 @@
 run:
+  modules-download-mode: readonly
   # Increase the default deadline from 1m as some module operations can take a
   # while if uncached!
   deadline: 5m

--- a/.golangci.yml
+++ b/.golangci.yml
@@ -1,6 +1,4 @@
 run:
-  modules-download-mode: readonly
-
   # Increase the default deadline from 1m as some module operations can take a
   # while if uncached!
   deadline: 5m

--- a/pkg/crd/markers/validation.go
+++ b/pkg/crd/markers/validation.go
@@ -26,6 +26,10 @@ import (
 	"sigs.k8s.io/controller-tools/pkg/markers"
 )
 
+const (
+	SchemalessName = "kubebuilder:validation:Schemaless"
+)
+
 // ValidationMarkers lists all available markers that affect CRD schema generation,
 // except for the few that don't make sense as type-level markers (see FieldOnlyMarkers).
 // All markers start with `+kubebuilder:validation:`, and continue with their type name.
@@ -82,6 +86,9 @@ var FieldOnlyMarkers = []*definitionWithHelp{
 
 	must(markers.MakeDefinition("kubebuilder:validation:EmbeddedResource", markers.DescribesField, XEmbeddedResource{})).
 		WithHelp(XEmbeddedResource{}.Help()),
+
+	must(markers.MakeDefinition(SchemalessName, markers.DescribesField, Schemaless{})).
+		WithHelp(Schemaless{}.Help()),
 }
 
 // ValidationIshMarkers are field-and-type markers that don't fall under the
@@ -224,6 +231,16 @@ type XPreserveUnknownFields struct{}
 // running apiserver. It is not necessary to add any additional schema for these
 // field, yet it is possible. This can be combined with PreserveUnknownFields.
 type XEmbeddedResource struct{}
+
+// +controllertools:marker:generateHelp:category="CRD validation"
+// Schemaless marks a field as being a schemaless object.
+//
+// Schemaless objects are not introspected, so you must provide
+// any type and validation information yourself. One use for this
+// tag is for embedding fields that hold JSONSchema typed objects.
+// Because this field disables all type checking, it is recommended
+// to be used only as a last resort.
+type Schemaless struct{}
 
 func (m Maximum) ApplyToSchema(schema *apiext.JSONSchemaProps) error {
 	if schema.Type != "integer" {

--- a/pkg/crd/markers/zz_generated.markerhelp.go
+++ b/pkg/crd/markers/zz_generated.markerhelp.go
@@ -306,6 +306,17 @@ func (Resource) Help() *markers.DefinitionHelp {
 	}
 }
 
+func (Schemaless) Help() *markers.DefinitionHelp {
+	return &markers.DefinitionHelp{
+		Category: "CRD validation",
+		DetailedHelp: markers.DetailedHelp{
+			Summary: "marks a field as being a schemaless object. ",
+			Details: "Schemaless objects are not introspected, so you must provide any type and validation information yourself. One use for this tag is for embedding fields that hold JSONSchema typed objects. Because this field disables all type checking, it is recommended to be used only as a last resort.",
+		},
+		FieldHelp: map[string]markers.DetailedHelp{},
+	}
+}
+
 func (SkipVersion) Help() *markers.DefinitionHelp {
 	return &markers.DefinitionHelp{
 		Category: "CRD",

--- a/pkg/crd/schema.go
+++ b/pkg/crd/schema.go
@@ -23,6 +23,7 @@ import (
 	"strings"
 
 	apiext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	crdmarkers "sigs.k8s.io/controller-tools/pkg/crd/markers"
 
 	"sigs.k8s.io/controller-tools/pkg/loader"
 	"sigs.k8s.io/controller-tools/pkg/markers"
@@ -378,7 +379,12 @@ func structToSchema(ctx *schemaContext, structType *ast.StructType) *apiext.JSON
 			}
 		}
 
-		propSchema := typeToSchema(ctx.ForInfo(&markers.TypeInfo{}), field.RawField.Type)
+		var propSchema *apiext.JSONSchemaProps
+		if field.Markers.Get(crdmarkers.SchemalessName) != nil {
+			propSchema = &apiext.JSONSchemaProps{}
+		} else {
+			propSchema = typeToSchema(ctx.ForInfo(&markers.TypeInfo{}), field.RawField.Type)
+		}
 		propSchema.Description = field.Doc
 
 		applyMarkers(ctx, field.Markers, propSchema, field.RawField)

--- a/pkg/crd/testdata/cronjob_types.go
+++ b/pkg/crd/testdata/cronjob_types.go
@@ -154,14 +154,19 @@ type CronJobSpec struct {
 
 	// This tests that min/max properties work
 	MinMaxProperties MinMaxObject `json:"minMaxProperties,omitempty"`
+
+	// This tests that the schemaless marker works
+	// +kubebuilder:validation:Schemaless
+	Schemaless []byte `json:"schemaless,omitempty"`
 }
 
 // +kubebuilder:validation:Type=object
 // +kubebuilder:pruning:PreserveUnknownFields
 type Preserved struct {
-	ConcreteField string `json:"concreteField"`
-	Rest map[string]interface{} `json:"-"`
+	ConcreteField string                 `json:"concreteField"`
+	Rest          map[string]interface{} `json:"-"`
 }
+
 func (p *Preserved) UnmarshalJSON(data []byte) error {
 	if err := json.Unmarshal(data, &p.Rest); err != nil {
 		return err

--- a/pkg/crd/testdata/testdata.kubebuilder.io_cronjobs.yaml
+++ b/pkg/crd/testdata/testdata.kubebuilder.io_cronjobs.yaml
@@ -5071,6 +5071,8 @@ spec:
               schedule:
                 description: The schedule in Cron format, see https://en.wikipedia.org/wiki/Cron.
                 type: string
+              schemaless:
+                description: This tests that the schemaless marker works
               startingDeadlineSeconds:
                 description: Optional deadline in seconds for starting the job if
                   it misses scheduled time for any reason.  Missed jobs executions

--- a/test.sh
+++ b/test.sh
@@ -107,12 +107,12 @@ fetch_kb_tools
 # setup testing env
 setup_envs
 
-header_text "running golangci-lint"
-
 header_text "generating marker help"
 pushd cmd/controller-gen > /dev/null
   go generate
 popd > /dev/null
+
+header_text "running golangci-lint"
 
 golangci-lint run --disable-all \
     --enable=misspell \

